pkg/ccl/streamingccl/streamingest/alter_replication_job.go

```diff
@@ -198,6 +219,91 @@ func alterReplicationJobHook(
 		if err != nil {
 			return err
 		}
+
+		// If a source address is being provided, we're enabling replication into an
+		// existing virtual cluster. It must be inactive, and we'll verify that it
+		// was the cluster from which the one it will replicate was replicated, i.e.
+		// that we're reversing the direction of replication. We will then revert it
+		// to the time they diverged and pick up from there.
+		if alterTenantStmt.ReplicationSourceAddress != nil {
+			dstTenantID, err := roachpb.MakeTenantID(tenInfo.ID)
+			if err != nil {
+				return err
+			}
+
+			// Here, we try to prevent the user from making a few
+			// mistakes. Starting a replication stream into an
+			// existing tenant requires both that it is offline and
+			// that it is consistent as of the provided timestamp.
+			if tenInfo.ServiceMode != mtinfopb.ServiceModeNone {
+				return errors.Newf("cannot start replication for tenant %q (%s) in service mode %s; service mode must be %s",
+					tenInfo.Name,
+					dstTenantID,
+					tenInfo.ServiceMode,
+					mtinfopb.ServiceModeNone,
+				)
+			}
+
+			streamAddress := streamingccl.StreamAddress(srcAddr)
+			streamURL, err := streamAddress.URL()
+			if err != nil {
+				return errors.Wrap(err, "url")
+			}
+			streamAddress = streamingccl.StreamAddress(streamURL.String())
+
+			client, err := streamclient.NewStreamClient(ctx, streamingccl.StreamAddress(srcAddr), p.ExecCfg().InternalDB)
+			if err != nil {
+				return errors.Wrap(err, "creating client")
+			}
+			srcID, resumeTS, err := client.PriorReplicationDetails(ctx, roachpb.TenantName(srcTenant))
+			if err != nil {
+				return errors.CombineErrors(errors.Wrap(err, "fetching prior replication details"), client.Close(ctx))
+			}
+			if err := client.Close(ctx); err != nil {
+				return err
+			}
+
+			if expected := fmt.Sprintf("%s:%s", p.ExtendedEvalContext().ClusterID, dstTenantID); srcID != expected {
+				return errors.Newf(
+					"tenant %q on specified cluster reports it was replicated from %q; %q cannot be rewound to start replication",
+					srcTenant, srcID, expected,
+				)
+			}
+
+			var revertFirst bool
+			if !tenInfo.LastRevertTenantTimestamp.Equal(resumeTS) {
+				revertFirst = true
```

Comment:
to clarify: `LastReverTenantTimestamp` == `resumeTS` if the user just ran cutover without starting service ? Could you add a small comment about that?

Reply:
I made it a const true.